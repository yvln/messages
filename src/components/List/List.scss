@import "styles/_variables.scss";

.List {
  > .Message {
    border-bottom: 1px solid $color-grey-light;

    &:last-child {
      border-bottom: none;
    }
  }
  
  > .text {
    padding: $spacing-medium;
    text-align: center;
  }
}

.List-blur {
  filter: blur(.2em);
}