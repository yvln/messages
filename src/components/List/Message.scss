@import "styles/_variables.scss";

.Message {
  padding: $spacing-medium;

  display: flex;
  flex-direction: column;
}

.Message-FirstLine {
  margin-bottom: $spacing-small;
  display: flex;
  justify-content: space-between;
}

.Message-FirstLine--text {
  > .username {
    font-weight: 600;
  }

  > .date {
    font-size: 12px;
    color: $color-grey;
  }
}

.Message-FirstLine--badge {
  text-transform: uppercase;
  text-align: center;
  font-size: 10px;

  width: 35px;
  padding: $spacing-small;

  border: 1px solid $color-grey-light;
  border-radius: .5em;
}