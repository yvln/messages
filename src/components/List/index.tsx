import { connect } from 'react-redux';

import * as actions from '../../modules/actions';
import { Dispatch, IFullState } from '../../types';
import List, { IDispatchProps, IStateProps } from './List';

const mapStateToProps = (state: IFullState): IStateProps => ({
  isFailure: state.messages.fetch.isFailure,
  isPending: state.messages.fetch.isPending,
  messages: state.messages.list,
});

const mapDispatchToProps = (dispatch: Dispatch): IDispatchProps => ({
  fetchMessages: () => {
    dispatch(actions.fetchMessages());
  },
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(List);
