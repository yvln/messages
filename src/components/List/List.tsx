import * as React from 'react';

import { IMessage } from '../../types';
import Container from '../Container';
import './List.scss';
import Message from './Message';

export interface IStateProps {
  isFailure: boolean;
  isPending: boolean;
  messages: IMessage[];
}

export interface IDispatchProps {
  fetchMessages: () => void;
}

interface IInjectedProps {
  username?: string;
}

interface IDefaultProps {
  messages: IMessage[];
}

export type IProps = IDefaultProps &
  IStateProps &
  IDispatchProps &
  IInjectedProps;

class List extends React.Component<IProps> {
  static defaultProps = {
    messages: [],
  };

  componentDidMount() {
    this.props.fetchMessages();
  }

  componentDidUpdate(prevProps: IProps) {
    if (prevProps.messages.length !== this.props.messages.length) {
      this.setState({
        messages: this.props.messages,
      });
    }
  }

  renderMessage = (message: IMessage) => {
    return <Message data={message} key={message.id.toString()} />;
  };

  renderList = () => {
    if (this.props.isFailure) {
      return (
        <div className="text">
          Unable to fetch messages. Please try again later.
        </div>
      );
    }

    if (!this.props.messages.length && !this.props.isPending) {
      return <div className="text">No messages yet.</div>;
    }

    if (!this.props.messages.length && this.props.isPending) {
      return <div className="text">...</div>;
    }

    return this.props.messages.map(this.renderMessage);
  };

  render() {
    const isToBlur = this.props.messages.length && this.props.isPending;

    return (
      <Container className={`List ${isToBlur ? 'List-blur' : ''}`}>
        {this.renderList()}
      </Container>
    );
  }
}

export default List;
