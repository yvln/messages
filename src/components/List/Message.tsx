import * as React from 'react';

import formatDate from '../../helpers/formatDate';
import { IMessage } from '../../types';
import './Message.scss';

interface IProps {
  key: string;
  data: IMessage;
}

class Message extends React.Component<IProps> {
  renderMessage = (): React.ReactNode => {
    const isPrivate = this.props.data.private;
    return (
      <>
        <div className="Message-FirstLine">
          <div className="Message-FirstLine--text">
            <span className="username">{this.props.data.username}</span> .{' '}
            <span className="date">{formatDate(this.props.data.date)}</span>
          </div>
          <div
            className={`Message-FirstLine--badge ${
              isPrivate ? 'Badge-private' : 'Badge-public'
            }`}
          >
            {isPrivate ? 'Private' : 'Public'}
          </div>
        </div>
        <div>{this.props.data.message}</div>
      </>
    );
  };

  render() {
    return (
      <div className="Message" key={this.props.data.id}>
        {this.renderMessage()}
      </div>
    );
  }
}

export default Message;
