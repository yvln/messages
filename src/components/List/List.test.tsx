import { configure, mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';

import List, { IProps } from './List';

configure({ adapter: new Adapter() });

describe('src/components/List', () => {
  let props: IProps;

  beforeEach(() => {
    props = {
      fetchMessages: jest.fn(),
      isFailure: false,
      isPending: false,
      messages: [
        {
          username: 'zoe',
          date: '2019-03-16T09:59:51.935Z',
          message: 'Duplexque isdem diebus acciderat malum, quod et',
          id: "sha1('2019-03-16T09:59:51.935Z-zoe')",
          private: false,
        },
      ],
    };
  });

  it('should render', () => {
    const component = shallow(<List {...props} />);
    expect(component).toMatchSnapshot();
  });

  it('should display a loader when is fetching and no messages yet.', () => {
    const component = mount(<List {...props} isPending={true} messages={[]} />);
    expect(component.text()).toContain('...');
  });

  it('should display an information when no messages and is not fetching', () => {
    const component = mount(<List {...props} messages={[]} />);
    expect(component.text()).toContain('No messages yet.');
  });

  it('should be blurred when is fetching but there are already messages', () => {
    const component = shallow(<List {...props} isPending={true} />);
    expect(component.hasClass('List-blur'));
  });
});
