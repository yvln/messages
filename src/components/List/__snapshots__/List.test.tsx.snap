// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/List should render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <List
    fetchMessages={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    isFailure={false}
    isPending={false}
    messages={
      Array [
        Object {
          "date": "2019-03-16T09:59:51.935Z",
          "id": "sha1('2019-03-16T09:59:51.935Z-zoe')",
          "message": "Duplexque isdem diebus acciderat malum, quod et",
          "private": false,
          "username": "zoe",
        },
      ]
    }
    username="yvln"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Message
          data={
            Object {
              "date": "2019-03-16T09:59:51.935Z",
              "id": "sha1('2019-03-16T09:59:51.935Z-zoe')",
              "message": "Duplexque isdem diebus acciderat malum, quod et",
              "private": false,
              "username": "zoe",
            }
          }
        />,
      ],
      "className": "List ",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "sha1('2019-03-16T09:59:51.935Z-zoe')",
        "nodeType": "class",
        "props": Object {
          "data": Object {
            "date": "2019-03-16T09:59:51.935Z",
            "id": "sha1('2019-03-16T09:59:51.935Z-zoe')",
            "message": "Duplexque isdem diebus acciderat malum, quod et",
            "private": false,
            "username": "zoe",
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Message
            data={
              Object {
                "date": "2019-03-16T09:59:51.935Z",
                "id": "sha1('2019-03-16T09:59:51.935Z-zoe')",
                "message": "Duplexque isdem diebus acciderat malum, quod et",
                "private": false,
                "username": "zoe",
              }
            }
          />,
        ],
        "className": "List ",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "sha1('2019-03-16T09:59:51.935Z-zoe')",
          "nodeType": "class",
          "props": Object {
            "data": Object {
              "date": "2019-03-16T09:59:51.935Z",
              "id": "sha1('2019-03-16T09:59:51.935Z-zoe')",
              "message": "Duplexque isdem diebus acciderat malum, quod et",
              "private": false,
              "username": "zoe",
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
